File: app/src/main/java/com/example/cn/helloworld/ui/playlist/PlaylistFragment.java
package com.example.cn.helloworld.ui.playlist;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.example.cn.helloworld.MusicActivity;
import com.example.cn.helloworld.R;
import com.example.cn.helloworld.data.model.Playlist;
import com.example.cn.helloworld.data.model.Song;
import com.example.cn.helloworld.data.playlist.PlaylistRepository;

import java.util.ArrayList;
import java.util.List;

public class PlaylistFragment extends Fragment {

    public static final String ARG_PLAYLIST_ID = "playlist_id";

    public static PlaylistFragment newInstance(String playlistId) {
        PlaylistFragment f = new PlaylistFragment();
        Bundle args = new Bundle();
        args.putString(ARG_PLAYLIST_ID, playlistId);
        f.setArguments(args);
        return f;
    }

    private PlaylistRepository playlistRepository;

    private ImageView coverImg;
    private TextView titleView;
    private TextView descView;
    private TextView tagView;
    private TextView metaView;
    private RecyclerView songList;

    private String playlistId;
    private SongAdapter songAdapter;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.fragment_playlist, container, false);

        playlistRepository = PlaylistRepository.getInstance(getContext());

        coverImg = (ImageView) view.findViewById(R.id.image_playlist_cover);
        titleView = (TextView) view.findViewById(R.id.text_playlist_title);
        descView = (TextView) view.findViewById(R.id.text_playlist_description);
        tagView = (TextView) view.findViewById(R.id.text_playlist_tags);
        metaView = (TextView) view.findViewById(R.id.text_playlist_meta);
        songList = (RecyclerView) view.findViewById(R.id.recycler_playlist_songs);   // ✔ 正确 ID

        songList.setLayoutManager(new LinearLayoutManager(getContext()));
        songAdapter = new SongAdapter(new ArrayList<Song>(), new SongAdapter.OnSongClickListener() {
            @Override
            public void onSongClick(Song song) {
                Context context = getContext();
                if (context == null || song == null || playlistId == null) {
                    return;
                }
                Intent intent = MusicActivity.createIntent(context, playlistId, song.getId());
                context.startActivity(intent);
            }
        });
        songList.setAdapter(songAdapter);

        loadPlaylist();

        return view;
    }

    private void loadPlaylist() {
        if (getArguments() == null) return;

        playlistId = getArguments().getString(ARG_PLAYLIST_ID);
        Playlist playlist = playlistRepository.getById(playlistId);
        if (playlist == null) return;

        // 封面
        if (playlist.getCoverUrl() != null && !playlist.getCoverUrl().isEmpty()) {
            coverImg.setImageURI(Uri.parse(playlist.getCoverUrl()));
        } else if (playlist.getCoverResId() != null) {
            coverImg.setImageResource(playlist.getCoverResId());
        }

        titleView.setText(playlist.getTitle());
        descView.setText(playlist.getDescription());

        // 标签
        if (playlist.getTags() != null && !playlist.getTags().isEmpty()) {
            tagView.setText(android.text.TextUtils.join(" / ", playlist.getTags()));
        }

        // 元信息：首数 / 播放量
        metaView.setText(
                playlist.getSongs().size() + " 首 · " + playlist.getPlayCount() + " 次播放"
        );

        songAdapter.setPlaylistCover(playlist.getCoverUrl(), playlist.getCoverResId());
        songAdapter.submit(playlist.getSongs());
    }


    // ================== Adapter ==================

    private static class SongAdapter extends RecyclerView.Adapter<SongAdapter.Holder> {

        interface OnSongClickListener {
            void onSongClick(Song song);
        }

        private List<Song> songs;
        private String playlistCoverUrl;
        private Integer playlistCoverResId;
        private final OnSongClickListener listener;

        SongAdapter(List<Song> s, OnSongClickListener listener) {
            songs = s;
            this.listener = listener;
        }

        void submit(List<Song> list) {
            songs = new ArrayList<>(list);
            notifyDataSetChanged();
        }

        void setPlaylistCover(String coverUrl, Integer coverResId) {
            this.playlistCoverUrl = coverUrl;
            this.playlistCoverResId = coverResId;
        }

        @NonNull
        @Override
        public Holder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View v = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_song, parent, false);
            return new Holder(v);
        }

        @Override
        public void onBindViewHolder(@NonNull Holder h, int pos) {
            final Song s = songs.get(pos);

            h.title.setText(s.getTitle());
            h.artist.setText(s.getArtist());
            h.duration.setText(formatDuration(s.getDurationMs()));
            h.description.setText(s.getDescription());

            bindCover(h.cover, s);
            h.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (listener != null) {
                        listener.onSongClick(s);
                    }
                }
            });
        }

        @Override
        public int getItemCount() {
            return songs.size();
        }

        static class Holder extends RecyclerView.ViewHolder {

            TextView title, artist, duration, description;
            ImageView cover;

            Holder(View v) {
                super(v);
                title = (TextView) v.findViewById(R.id.text_song_title);
                artist = (TextView) v.findViewById(R.id.text_song_artist);
                duration = (TextView) v.findViewById(R.id.text_song_duration);
                description = (TextView) v.findViewById(R.id.text_song_description);
                cover = (ImageView) v.findViewById(R.id.image_song_cover);
            }
        }

        private void bindCover(ImageView coverView, Song song) {
            if (coverView == null || song == null) {
                return;
            }

            if (!TextUtils.isEmpty(song.getCoverImagePath())) {
                coverView.setImageURI(Uri.parse(song.getCoverImagePath()));
            } else if (!TextUtils.isEmpty(song.getCoverUrl())) {
                coverView.setImageURI(Uri.parse(song.getCoverUrl()));
            } else if (song.getCoverResId() != 0) {
                coverView.setImageResource(song.getCoverResId());
            } else if (!TextUtils.isEmpty(playlistCoverUrl)) {
                coverView.setImageURI(Uri.parse(playlistCoverUrl));
            } else if (playlistCoverResId != null) {
                coverView.setImageResource(playlistCoverResId);
            } else {
                coverView.setImageResource(R.drawable.cover_playlist_placeholder);
            }
        }

        private static String formatDuration(long durationMs) {
            long totalSeconds = durationMs / 1000;
            long minutes = totalSeconds / 60;
            long seconds = totalSeconds % 60;
            return String.format(java.util.Locale.getDefault(), "%02d:%02d", minutes, seconds);
        }
    }
}

File: app/src/main/java/com/example/cn/helloworld/ui/playlist/PlaylistFragment.kt
package com.example.cn.helloworld.ui.playlist

import android.arch.lifecycle.Lifecycle
import android.arch.lifecycle.LifecycleOwner
import android.arch.lifecycle.LifecycleRegistry
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.util.DiffUtil
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import com.example.cn.helloworld.R
import com.example.cn.helloworld.data.model.Playlist
import com.example.cn.helloworld.data.model.Song
import java.util.Locale

class PlaylistFragment : Fragment(), LifecycleOwner {

    private val lifecycleRegistry = LifecycleRegistry(this)

    private lateinit var playlistCover: ImageView
    private lateinit var playlistTitle: TextView
    private lateinit var playlistDescription: TextView
    private lateinit var playlistMeta: TextView
    private lateinit var songsRecyclerView: RecyclerView
    private lateinit var songsAdapter: PlaylistSongsAdapter

    private lateinit var viewModel: PlaylistViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_CREATE)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_playlist, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        playlistCover = view.findViewById(R.id.image_playlist_cover)
        playlistTitle = view.findViewById(R.id.text_playlist_title)
        playlistDescription = view.findViewById(R.id.text_playlist_description)
        playlistMeta = view.findViewById(R.id.text_playlist_meta)
        songsRecyclerView = view.findViewById(R.id.recycler_playlist_songs)

        songsAdapter = PlaylistSongsAdapter { song ->
            val context = view.context
            Toast.makeText(context, "准备播放：${song.title}", Toast.LENGTH_SHORT).show()
        }

        songsRecyclerView.apply {
            layoutManager = LinearLayoutManager(context)
            adapter = songsAdapter
            setHasFixedSize(true)
        }

        viewModel = ViewModelProviders.of(this).get(PlaylistViewModel::class.java)
        viewModel.playlist.observe(this, Observer { playlist ->
            if (playlist != null) {
                bindPlaylist(playlist)
            }
        })
        viewModel.songs.observe(this, Observer { songs ->
            songsAdapter.submitList(songs)
        })
    }

    override fun onStart() {
        super.onStart()
        lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_START)
    }

    override fun onResume() {
        super.onResume()
        lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_RESUME)
    }

    override fun onPause() {
        lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_PAUSE)
        super.onPause()
    }

    override fun onStop() {
        lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_STOP)
        super.onStop()
    }

    override fun onDestroy() {
        lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_DESTROY)
        super.onDestroy()
    }

    override fun getLifecycle(): Lifecycle = lifecycleRegistry

    private fun bindPlaylist(playlist: Playlist) {
        playlistTitle.text = playlist.title
        playlistDescription.text = playlist.description
        playlistMeta.text = buildMeta(playlist)

        val coverResId = playlist.coverResId
        if (coverResId != null) {
            playlistCover.setImageResource(coverResId)
        } else {
            playlistCover.setImageResource(R.drawable.cover_lisao)
        }
    }

    private fun buildMeta(playlist: Playlist): String {
        val count = playlist.songs.size
        val totalDurationMs = playlist.songs.sumByDouble { it.durationMs.toDouble() }
        val totalMinutes = (totalDurationMs / 1000 / 60).toInt()
        val tags = if (playlist.tags.isNotEmpty()) playlist.tags.joinToString(" / ") else ""
        return when {
            tags.isNotEmpty() -> String.format(Locale.getDefault(), "共 %d 首 · %d 分钟 · %s", count, totalMinutes, tags)
            else -> String.format(Locale.getDefault(), "共 %d 首 · %d 分钟", count, totalMinutes)
        }
    }

    private class PlaylistSongsAdapter(
        private val onItemClick: (Song) -> Unit
    ) : SimpleListAdapter<Song, PlaylistSongsAdapter.SongViewHolder>(SongDiffCallback()) {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SongViewHolder {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.item_song, parent, false)
            return SongViewHolder(view, onItemClick)
        }

        override fun onBindViewHolder(holder: SongViewHolder, position: Int) {
            holder.bind(getItem(position))
        }

        private class SongViewHolder(
            itemView: View,
            private val onItemClick: (Song) -> Unit
        ) : RecyclerView.ViewHolder(itemView) {

            private val coverView: ImageView = itemView.findViewById(R.id.image_song_cover)
            private val titleView: TextView = itemView.findViewById(R.id.text_song_title)
            private val artistView: TextView = itemView.findViewById(R.id.text_song_artist)
            private val durationView: TextView = itemView.findViewById(R.id.text_song_duration)
            private val descriptionView: TextView = itemView.findViewById(R.id.text_song_description)

            fun bind(song: Song) {
                titleView.text = song.title
                artistView.text = song.artist
                durationView.text = formatDuration(song.durationMs)

                val description = song.description
                if (description.isNullOrEmpty()) {
                    descriptionView.visibility = View.GONE
                } else {
                    descriptionView.visibility = View.VISIBLE
                    descriptionView.text = description
                }

                val coverRes = song.coverResId
                if (coverRes != null) {
                    coverView.setImageResource(coverRes)
                } else {
                    coverView.setImageResource(R.drawable.cover_baobei)
                }

                itemView.setOnClickListener { onItemClick(song) }
            }

            private fun formatDuration(durationMs: Long): String {
                val totalSeconds = durationMs / 1000
                val minutes = totalSeconds / 60
                val seconds = totalSeconds % 60
                return String.format(Locale.getDefault(), "%02d:%02d", minutes, seconds)
            }
        }

        private class SongDiffCallback : DiffUtil.ItemCallback<Song>() {
            override fun areItemsTheSame(oldItem: Song, newItem: Song): Boolean = oldItem.id == newItem.id

            override fun areContentsTheSame(oldItem: Song, newItem: Song): Boolean = oldItem == newItem
        }
    }

    private abstract class SimpleListAdapter<T, VH : RecyclerView.ViewHolder>(
        private val diffCallback: DiffUtil.ItemCallback<T>
    ) : RecyclerView.Adapter<VH>() {

        private val items = mutableListOf<T>()

        protected fun getItem(position: Int): T = items[position]

        override fun getItemCount(): Int = items.size

        fun submitList(newItems: List<T>?) {
            val targetItems = newItems ?: emptyList()
            val diffResult = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
                override fun getOldListSize(): Int = items.size

                override fun getNewListSize(): Int = targetItems.size

                override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    return diffCallback.areItemsTheSame(
                        items[oldItemPosition],
                        targetItems[newItemPosition]
                    )
                }

                override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    return diffCallback.areContentsTheSame(
                        items[oldItemPosition],
                        targetItems[newItemPosition]
                    )
                }
            })
            items.clear()
            items.addAll(targetItems)
            diffResult.dispatchUpdatesTo(this)
        }
    }

    companion object {
        fun newInstance(): PlaylistFragment = PlaylistFragment()
    }
}
